### PR TITLE
python3: Add C++ compiler support for distutils (on non gcc systems)

### LINKED ISSUE #18194: pythonPackages.gdal broken on x86_64-darwin
## Issue description

This package is reporting that it is not supported on darwin systems. I discovered this issue when trying to update `pythonPackages.ansible2`, which had worked cleanly previously (I last updated probably a month ago). I looked through the nixpkgs Git log to see if any recent _gdal_ changes were pushed related to OS X specifically, but I did not see anything too obvious. Perhaps this commit 674aa05abca627d?
### Steps to reproduce

```
$ nix-env -iA nixpkgs.gdal
error: Package ‘gdal-2.1.1’ in ‘/nix/store/3ix4qwxv1l4czrsb4ipy2wv08pxrs7mh-nixpkgs-16.09pre90351.e446b9f/nixpkgs/pkgs/development/libraries/gdal/default.nix:59’ is not supported on ‘x86_64-darwin’, refusing to evaluate.
a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowBroken = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build` or any other Nix command you can add
  { allowBroken = true; }
to ~/.nixpkgs/config.nix.
```
## Technical details
- System: OS X El Capitan (10.11.6)
- Nix version: nix-env (Nix) 1.11.3
- Nixpkgs version: "16.09pre-git" (nixpkgs-unstable channel updated on 2016-09-01)

### PR DESCRIPTION
**Pull request by knedlsepp**
###### Motivation for this change
Python 3.x on darwin did not correctly distinguish between C++ and C builds, so we got errors relating to missing C++ headers. The fix is analogous to the one for python2.7 done in #19585.
This fixes #18194.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by siddharthist** [comment]
This indeed fixes the build of `gdal` and so `django` on OSX with Python 3.6 😄  🎆 .

**Comment by FRidh** [comment]
What bugs.python.org issue is this patch linked from?

**Comment by knedlsepp** [comment]
It's from http://bugs.python.org/issue1222585 should I add that in the commit messages?

**Comment by FRidh** [comment]
@knedlsepp in a comment in the code please. Also, what about http://bugs.python.org/issue1222585#msg242706?

**Comment by knedlsepp** [comment]
@FRidh: Thank you for the thorough look at this. This is something that I overlooked and I need to incorporate before we merge!

File: pkgs/development/interpreters/python/cpython/3.6/default.nix
Comment (LnL7) [inline pkgs/development/interpreters/python/cpython/3.6/default.nix]: Is there a reason not to apply this for gcc?

Comment (knedlsepp) [inline pkgs/development/interpreters/python/cpython/3.6/default.nix]: Well. The problem does not occur when using gcc, so I thought it would be best to make the intention clear. I'm not sure if applying this unconditionally might break something else, so I am basically being cautious.

**Comment by knedlsepp** [comment]
Please do not merge ATM, just broke it.

**Comment by knedlsepp** [comment]
@FRidh: I patched the patch based on the comment you mentioned and did upload it to their issue tracker. My reworked PR now links to the patch I uploaded. I also put a comment in the code mentioning the issue. Did a little test with compiling a C++ `BOOST_PYTHON_MODULE`, which looks fine for 3.4, 3.5 and 3.6.

File: pkgs/development/interpreters/python/cpython/3.6/default.nix
Comment (LnL7) [inline pkgs/development/interpreters/python/cpython/3.6/default.nix]: How I understand it is that it currently only works because gcc doesn't complain when compiling c++ code using `$CC`, clang does which is arguably the correct thing to do 😄 .
In cases like this I prefer not to use conditionals, making everything more consistent and the darwin build less brittle. (if the patch fails to apply the linux build won't break)

Comment (knedlsepp) [inline pkgs/development/interpreters/python/cpython/3.6/default.nix]: Ok. In that case I should probably boot up a little NixOS VM and try if that this also works fine. BTW: The python27 currently also does this only on !isGNU, should we maybe change that?

**Comment by FRidh** [comment]
@knedlsepp thanks! Is there anything that should be done with the 2.7 version of the patch?

File: pkgs/development/interpreters/python/cpython/3.6/default.nix
Comment (knedlsepp) [inline pkgs/development/interpreters/python/cpython/3.6/default.nix]: @FRidh Just saw your post on the 2.7 version of this issue: https://github.com/NixOS/nixpkgs/pull/19585#discussion_r83555826 what's your take on `!isGNU`?

Comment (FRidh) [inline pkgs/development/interpreters/python/cpython/3.6/default.nix]: @LnL7 has a valid point but I would prefer to keep the Python interpreter as vanilla as possible and therefore only apply patch when really necessary.

**Comment by knedlsepp** [comment]
@FRidh Hm, we might: https://github.com/NixOS/nixpkgs/pull/19585#discussion_r83694949

File: pkgs/development/interpreters/python/cpython/3.6/default.nix
Comment (copumpkin) [inline pkgs/development/interpreters/python/cpython/3.6/default.nix]: I generally prefer to minimize amount of "branchy" logic between platforms, if we can pull it off. I'd make this unconditional, adding a comment saying what we say here, i.e., that gcc is okay with it and clang isn't, and we're doing this so we don't bake in assumptions about weird behavior in specific compilers.

**Comment by knedlsepp** [comment]
@FRidh: Could we merge this? While there might still be some open issues considering dropping the CFLAGS on 2.7, I think we should nevertheless get 3.x up and running.

**Comment by FRidh** [comment]
This will apply the patch unconditionally unlike your other patch.

`++ optionals (!(stdenv.cc.isGNU or false))` ?

**Comment by knedlsepp** [comment]
I understand your point of wanting to keep the interpreter as vanilla as possible.  As the motion in https://github.com/NixOS/nixpkgs/pull/27649#pullrequestreview-52419327 however was mostly: Applying the patch unconditionally will be less painful for darwin maintenance, I went with that approach instead. I think the patch failing to apply for a maintainer only testing on linux could be seen as a useful information. The 2.7-patch was actually incorporated by veprbl, I only copied the fix.

**Comment by veprbl** [comment]
Have you tested numpy with this?
There is a bit of patching there that you don't seem to engage:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/numpy/default.nix#L16-L20

**Review by matthewbauer — CHANGES_REQUESTED** [review]
Merge conflict

**Comment by knedlsepp** [comment]
Closing in favor of: #39576

**Pull request closed without merging**
